### PR TITLE
lib: make extensible

### PR DESCRIPTION
**Pull request by infinisil**
This allows the lib fixed point to be extended with

```nix
myLib = lib.extend (self: super: {
  foo = "foo";
})
```

With this it's possible to have the new modified lib attrset available to all
modules when using evalModules

```nix
myLib.evalModules {
  modules = [ ({ lib, ... }: {
    options.bar = lib.mkOption {
      default = lib.foo;
    };
  }) ];
}
```
=>
```nix
{ config = { bar = "foo"; ... }; options = ...; }
```

###### Motivation for this change
This will be useful for projects like @rycee [home-manager](https://github.com/rycee/home-manager) that use the module system themselves without the ability to modify `nixpkgs/lib` directly. Currently home-manager needs to [use](https://github.com/rycee/home-manager/blob/7c9278bd92a067832c421e79f25401a9791aaddd/modules/modules.nix#L93) the [prone to infinite recursion](https://github.com/rycee/home-manager/pull/160) lib module that just adds a `lib` option. Stuff like `with config.lib;` at the top of the file will give you infinite recursion in general though.

Another instance is @LnL7's [nix-darwin](https://github.com/LnL7/nix-darwin) which currently [imports its lib directly](https://github.com/LnL7/nix-darwin/blob/master/modules/launchd/default.nix#L3) in every file that needs it (which is what home-manager used previously).

@LnL7 @rycee @Profpatsch @edolstra @shlevy 

###### Things done

Other than running a few very basic commands like the one above I haven't tested this very well. But I'm 99% sure this can't break anything and is fully backwards compatible. Will test it more in a bit.

### CODE REVIEW

File: lib/default.nix [hunk 1]
```diff
@@ -5,13 +5,16 @@
...
+  lib = fixedPoints.makeExtensible (self: let
+    callLibs = file: import file { lib = self; };
+  in rec {
 
-  lib = rec {
 
     # often used, or depending on very little
     trivial = callLibs ./trivial.nix;
-    fixedPoints = callLibs ./fixed-points.nix;
+    inherit fixedPoints;
```
Comment (Profpatsch) [inline lib/default.nix]: I’d keep the symmetry with other calls in this module.

File: lib/default.nix [hunk 2]
```diff
@@ -5,13 +5,16 @@
  */
 let
 
-  callLibs = file: import file { inherit lib; };
+  fixedPoints = import ./fixed-points.nix {};
+
+  lib = fixedPoints.makeExtensible (self: let
+    callLibs = file: import file { lib = self; };
+  in rec {
```
Comment (Profpatsch) [inline lib/default.nix]: The indentation is broken from here.

**Review by Profpatsch — COMMENTED** [review]
I’m not sure this is the correct way to implement overriding `lib`.
@edolstra, @nbp and maybe @shlevy and @aszlig should take a look.

File: lib/default.nix (hunk 2 above)
Comment (Infinisil) [reply lib/default.nix]: How would you format it?

Comment (Ericson2314) [reply lib/default.nix]: Don't use `rec`. Use `with self;` instead.

**Comment by Infinisil** [comment]
Changed it to `with self;`

File: lib/default.nix [hunk 3]
```diff
@@ -5,13 +5,16 @@
  */
 let
 
-  callLibs = file: import file { inherit lib; };
+  fixedPoints = import ./fixed-points.nix {};
```
Comment (Ericson2314) [inline lib/default.nix]: Do `inherit (import ./fixed-points.nix {}) makeExtensible!`.

**Review by Ericson2314 — COMMENTED** [review]
Fix that and @Profpatsch's comment on symmetry, and this looks good to me. Thanks!

Sorry to not catch that before.

**Comment by Ericson2314** [comment]
@Profpatsch with our changes this does look indeed look correct to me.

File: lib/default.nix (hunk 3 above)
Comment (Infinisil) [reply lib/default.nix]: Ah yeah that sounds better

**Comment by Infinisil** [comment]
Updated

File: lib/default.nix [hunk 4]
```diff
@@ -5,13 +5,16 @@
...
 
-  lib = rec {
+  lib = makeExtensible (self: let
+    callLibs = file: import file { lib = self; };
+  in with self; {
 
     # often used, or depending on very little
     trivial = callLibs ./trivial.nix;
     fixedPoints = callLibs ./fixed-points.nix;
+    inherit fixedPoints;
```
Comment (Ericson2314) [inline lib/default.nix]: Ah this line still needs to go.

**Review by Ericson2314 — COMMENTED** [review]
Ah forgot to remove old inherit.

File: lib/default.nix (hunk 4 above)
Comment (Infinisil) [reply lib/default.nix]: I excuse my sluggishness

**Comment by Ericson2314** [comment]
Great! I'll merge after of borg.

**Comment by grahamc** [comment]
I'm not sure lib should be so easy to extend, since it sort of obscures what is in lib, and possible collisions when new functions are added later on.

**Comment by zimbatm** [comment]
I was concerned about evaluation time. result: it doesn't affect `time nix-env -qa >/dev/null` (around 6s on my laptop).

**Comment by Infinisil** [comment]
@grahamc The same could be said for pkgs overlays though, but these are doing just fine with a much larger attribute count. Also this is no problem of nixpkgs, because extending lib is mostly useful outside of it. If you need to extend lib within nixpkgs you can just add the function you need in nixpkgs/lib as it's been done all this time.

**Comment by grahamc** [comment]
I guess I don't see the benefit of making it extensible, but I can see the benefits of keeping it closed -- preferring explicitness in what is contained in the nixpkgs lib. pkgs being an extensible fixed point isn't a good comparison, because the fixed point is required for dependency resolution and replacing packages -- a desirable set of features which I don't think translate in to eval-time library code.

**Comment by zimbatm** [comment]
@Infinisil why can't the overlay have a `myoverlay.lib` key with their own extensions?

**Comment by shlevy** [comment]
I'm with @grahamc here. I think we need concrete use cases to add more configuration knobs and increase complexity, and I just don't see why we need this. Note that we already have `config.lib` in NixOS to allow modules to give each other helper functions...

There is no real analogy with package set extensibility IMO.

**Comment by shlevy** [comment]
Re infinite recursion, we have `mkIf` etc. already there.

**Comment by Infinisil** [comment]
An alternative, while not as elegant, is to add some functionality to the module system that allows you to set your own lib.

**Review by nbp — COMMENTED** [review]
The changes made in `lib/modules.nix` sounds fine to me.

However, I will note that unless you are extending the module system you should probably not do that. 

Instead you can set the `_module.args.hmLib` attribute to whatever new set of library function you want to be added. The reason of having a `lib` attribute is only for getting the module system library functions without the `pkgs` attribute.

**Comment by rycee** [comment]
Since this (apparently) has no negative impact on today's use in NixOS and Nixpkgs I'd say that its an interesting change that might be well worth experimenting with in non-Nixpkgs projects that use the NixOS module system. Of course, I'm personally thinking about Home Manager here.

I'm also liking nbp's idea of adding a new attribute for the library set but would like the ability to test both variants.

I should also add that I'm partial to this change also on purely esthetic grounds; the symmetry it adds is quite nice :-)

**Comment by Infinisil** [comment]
Also, this change allows the module system itself to be modified/extended, which would've not been possible without forking nixpkgs before. And I think it makes sense to enable users of nixpkgs to do that. And as said by @rycee, I don't see any negative impact on nixpkgs/NixOS with this.

**Comment by Profpatsch** [comment]
> Instead you can set the _module.args.hmLib attribute to whatever new set of library function you want to be added. The reason of having a lib attribute is only for getting the module system library functions without the pkgs attribute.

Yes, I’d prefer that to making `lib` extensible.

**Comment by Infinisil** [comment]
I discussed other options for modifying the `lib` argument with @nbp on #nixos and the best thing we could come up with is to use `evalModules` `specialArgs` argument, see https://gist.github.com/Infinisil/aa97115ccec082ad192088b17b92f011#gistcomment-2427130. While  specialArgs overriding normal arguments like `config`, `options` and `lib` doesn't really look like intended behaviour, it's what makes this work. While it's not as nice as the extensible lib, I'd almost be okay with using this instead. I really don't want all modules not in nixpkgs to have to use
```nix
{ lib, myLib, ... }:
with lib;
with myLib;
```

**Comment by Profpatsch** [comment]
> I really don't want all modules not in nixpkgs to have to use

What’s so bad about that?

**Comment by Infinisil** [comment]
@Profpatsch If I didn't know the internals of nixpkgs and saw that code I'd ask myself why there are two libs and why they couldn't be combined

**Comment by nbp** [comment]
One last thing, using `specialArgs` or an extended `lib` (independently of how it is extended) will prevent using the home-manager as a sub-module, which I think is desirable for use cases where the home manager is used to configure some NixOS users.

Therefore, we should find a solution for the `with` problem described in https://github.com/NixOS/nixpkgs/pull/38522#issuecomment-379584931

**Comment by Infinisil** [comment]
Now it's been some time since this was (a bit prematurely) merged and while there were a lot of voices against it, I feel like nobody's gonna undo the change themselves. While I'd love to use this new functionality, I'm gonna hold back until I know this is gonna stay.

To get an idea of amount of agreement I'm asking everybody here to **vote on this comment** with
- 👍 You like this change and don't want this to be reverted
- 👎 You dislike this change, nixpkgs is better off without it and it should be reverted.

**Comment by oxij** [comment]
My two cents.

At first I was somewhere between "meh, whatever" to :-1: on this. But after thinking about this for a while, now I feel like nixpkgs should follow GHC's approach of "it's always ok to remove a limitation that has no technical justification". Hence, now I'm completely :+1: on this until somebody can demonstrate technical, not stylistic, justifications against this.

I feel like if you want this to be reverted it's only fair to ask for overlays to be reverted too as they add a similar amount of complexity.

**Comment by shlevy** [comment]
Again, there is no real comparison to overlays here. There have been use cases with a strong need to modify or add to the base package set for years before overlays, and they have been served very well since overlays were introduced. On the other hand, I've literally never felt the need to either override or extend `lib`, nor seen a compelling example of its usefulness.

**Comment by oxij** [comment]
I get your stylistic argument, overriding lib is uncommon. But what is the technical argument against this?

And, for the record, I do extend the `lib` (with stuff I should probably just PR, but let's assume it's very my-configs specific for the sake of the argument).
Can I just import a file whenever I use those things? Sure. Can I use module arguments? Sure.

But what if I want to use those functions in package expressions in my overlay?
Can I just import a file there? Well, sure, but what is the technical difference with overlays, again?

Personally, I don't even use overlays, I just dump everything into branches (because I prefer to fix merge conflicts as opposed to debugging why things unexpectedly silently broke with overlays), but if I did use them for my overrides I would see the same benefits with this PR.

**Comment by oxij** [comment]
Actually, I think I just came up with a use for this (or a bit modified version of this) directly in nixpkgs.

Why not make `nixos/lib/*` into such an extension of the `lib`? That would save a lot of imports in nixos sources. Some duplicated code would be easier to move there too if it were a lib (I wanted to implement `nixoslib` or something eventually, but something like this seems fine too). And you could still evaluate `nixpkgs` against the default `lib` as before with some minimal modifications.

**Comment by Infinisil** [comment]
@oxij Interesting, very good point. Some excerpts from nixpkgs/lib:

https://github.com/NixOS/nixpkgs/blob/12ce0db1bffb048fbf81a852aed7240aedc31757/nixos/lib/build-vms.nix#L1-L6

https://github.com/NixOS/nixpkgs/blob/12ce0db1bffb048fbf81a852aed7240aedc31757/nixos/lib/eval-config.nix#L25-L27

This is the whole file:
https://github.com/NixOS/nixpkgs/blob/12ce0db1bffb048fbf81a852aed7240aedc31757/nixos/lib/from-env.nix#L1-L4

https://github.com/NixOS/nixpkgs/blob/12ce0db1bffb048fbf81a852aed7240aedc31757/nixos/lib/utils.nix#L1-L3

I think all of these could be implemented by extending lib. I'm not 100% sure how to handle the `pkgs` argument though, since that's not really a thing in `nixpkgs/lib`.

**Pull request merged**